### PR TITLE
Upgrade to Conda Build 3 (finally) and Move to Azure DevOps

### LINKED ISSUE #125: Change source docker image to Omnia from Jchodera
We have been uploading and getting Docker images from @jchodera's Docker Hub profile, however, that is not tenable long term as we will want multiple admins to work on the docker containers going forward. 

To that end, @jchodera and I have created the Omnia MD organization on [Docker Hub](https://hub.docker.com/r/omniamd/omnia-linux-anvil/) (+ [Docker Cloud](https://cloud.docker.com/swarm/omniamd/repository/docker/omniamd/omnia-linux-anvil/general) so multiple people can be admins and control the builds in the future. 

We will eventually need to change the [`.travis.yml` line](https://github.com/omnia-md/conda-dev-recipes/blob/master/.travis.yml#L59) from  `jchodera/omnia-linux-anvil:{tag}` to `omniamd/omnia-linux-anvil:{tag}`. Right now both accounts are building the same tags starting with `cuda91` as we are debugging the new image and the migration, so this is an "eventually" action.

### LINKED ISSUE #160: Update travis osx build from xcode 6.4
Travis support xcode 6.4 is [going away in Jan 2019](https://github.com/travis-ci/travis-web/pull/1759/commits) (this month!):
>  Running builds with Xcode 6.4 in Travis CI is deprecated and will be removed in January 2019. If Xcode 6.4 is critical to your builds, please contact our support team at support@travis-ci.com to discuss options.

We'll need to update our builds to use a more modern xcode.

cc: https://github.com/omnia-md/conda-recipes/issues/928

### LINKED ISSUE #176: Uploads are broken
@peastman: Just wanted to point out that @j-wags noticed that omnia builds are broken again, with the last successful upload being 12 days ago.

I note that the last successful `omnia-dev` nightly upload of `openmm` was 24 days ago:
https://anaconda.org/omnia-dev/openmm/files

Did you fix anything to get builds working again for https://github.com/omnia-md/conda-recipes-cuda? If so, could you make those changes here too?

### PR DESCRIPTION
**Pull request by Lnaden**
After a bunch of testing, this is the working (at least in dry runs) upgrade to the Omnia conda-build infrastructure. This replaces #154 and does the following things:

* Total overhaul of the Conda-Build-All script to use Conda Build 3+ API
* Drops Travis and AppVeyor and replaces with Azure DevOps
* Can be integrated into conda-recipes repo with the goal of rendering this obsolete (topic for discussion)
* Nightly Builds established through same YAML as non-nightly builds
* Uses our new docker images from the `omniamd` Docker Group for linux
* Splits the Python and CUDA builds 1 per combination of Python {2.7, 3.7, 3.8} and CUDA {8.0, 9.0, 9.1, 9.2, 10.0, 10.1, 10.2}. This results in a large number of builds, but each one is specialized at least and we don't have to fight timeout.
* The ability to ignore and include explicit Jinja keys in the recipe

You can see the output CI stuff here: https://dev.azure.com/lnaden/conda-dev-recipes/_build/results?buildId=72

### Outstanding Issues 
* I'm still using the `--dry-run` flag and have not adjusted the OpenMM `build.sh` and `build.bat` scripts. 
* I am currently only building OpenMM in the CUDA builds through the jinja include/exclude logic I added. However, how do we want to handle OpenMM builds without CUDA?
* What do we want to call the OpenMM builds which are not built with CUDA
* I need to actually test a proper build and upload with the script, although the majority of that logic should not have changed.
* we will need to transfer the BINSTAR_TOKEN to Azure from Travis and/or AppVeyor

### Theoretical Issue closures

* Closes #176
* Closes #175
* Closes #167
* Closes #161
* Closes #160
* Closes #125
* Closes #111 
* Closes #59

### CODE REVIEW

**Comment by Lnaden** [comment]
And cc @jchodera @j-wags @peastman because I'll need your input to finish. At this point, it should be working from a technical stand point, we just need to finalize all the build flags and conditions.

**Comment by Lnaden** [comment]
why.. why did that close it? It should stay open

**Comment by Lnaden** [comment]
@jaimergp Continuing our conversation (this is just copy from the slack):

> Whatever the default main label is listing, but it shouldn’t matter anyway.

There is one problem I see with this is that we will still need to build a osx version since the current setup I have is OpenMM building only on CUDA images/installs.

**Comment by Lnaden** [comment]
To that end, we will need to figure out a build string for OpenMM which works with or without the `CUDA_SHORT_VERSION` jinja being set.

**Comment by jaimergp** [comment]
Can we use `nocuda` when CUDA_SHORT_VERSION is not set?

**Comment by Lnaden** [comment]
Sure, I just need to figure out the Jinja, and we still need to make sure that when someone does `conda install openmm` they get the correct version.

**Comment by Lnaden** [comment]
here is the syntax I assembled:

```
{% if CUDA_SHORT_VERSION is defined %}
{% set CUDA_STR = "cuda" + CUDA_SHORT_VERSION %}
{% else %}
{% set CUDA_STR = "nocuda" %}
{% endif %}
```

Can then use: `string: py{{ py }}_{{ CUDA_STR }}_{{ PKG_BUILDNUM }}` later

**Comment by jaimergp** [comment]
Do we want to add the package hash (`{{ PKG_HASH }}`) too? See [`numpy` at `conda-forge` for an example](https://anaconda.org/conda-forge/numpy/files).

**Comment by Lnaden** [comment]
> Do we want to add the package hash ({{ PKG_HASH }}) too?

My first thought is "no" for the reason that I explicitly did not want to re-build existing packages which don't need an update. The "Does the package need to be built" checks the output tarball name against the one on the Conda CDN's and skips unless forced. This worked until Conda-build started hashing the dependency pinning in Conda-Build 3.something. To keep that behavior, I intentionally added the `filename_hashing=False` kwarg in the `conda_build.api.render` call which is equivalent to the CLI `conda build --old-build-string`.

However, since we are now dropping 2.7, we need to build for 3.8, and I'm pretty sure most of the 3.7 builds are missing anyways, it may be worth dropping that and letting packages build with the hashes now once again. We would still have to manually add it to the `string` for OpenMM since we are manipulating it anyways, but we could! 

tl;dr: maybe the answer should be "yes?" Thoughts?

**Comment by peastman** [comment]
Would it be difficult to also create builds for Python 3.6?  I know that's one more version than conda-forge builds for, but we've always tried to support the three most recent versions, and I think that's a more realistic reflection of how quickly users upgrade.  For example, take a look at the download statistics for OpenMM 7.4 (released less than three months ago).  The Mac version has had 2441 downloads for Python 3.7, 2350 downloads for 3.6, and 256 downloads for 3.5.  The ratios for Linux downloads are similar.

**Comment by Lnaden** [comment]
> Would it be difficult to also create builds for Python 3.6?

Not difficult, it just adds more builds in this setup. Its as easy as adding more 4-line YAML conda config files and extending the Azure ones. 

>  but we've always tried to support the three most recent versions,

I dont think thats quite true, I seem to only remember supporting the 2 most recent and we faded out the older version as we got all the new builds fixed. This was also back when we had to fix the build system every time there was a new Python version, so that may have been a stronger factor.

**Comment by Lnaden** [comment]
I'm fine leaving them in, it just adds more build time is all

**Comment by peastman** [comment]
Is build time a problem?  We only do a few releases a year, so it shouldn't be a big deal.  We can do fewer Python versions for nightly builds, if those are an issue.

**Comment by Lnaden** [comment]
> Is build time a problem?

 not really. I am going to port this to the main omnia repo too. ideally get all our builds in the same location

**Comment by jchodera** [comment]
Apologies for not being able to review this earlier! Reviewing this now, since it looks we need to put this into production.

**Comment by jchodera** [comment]
Actually, @jaimergp and I will meet about this tomorrow morning and should be able to review this then!

**Review by jchodera — APPROVED** [review]
@Lnaden : Thanks so much for doing this! @jaimergp have gone through this.

Here is a summary of our discussion for consideration in future updates:
* **Supported Python versions:** `conda-forge` currently officially supports [Python 2.7, 3.6 and 3.7](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml#L557), so we should try to support these if possible. They do have a [bot adding 3.8 support](https://conda-forge.org/status/), so we should add that as well. I'm certainly in favor of dropping Python 2.7 ASAP, but it would be best if we can keep in sync with what conda-forge supports.
* **Azure Devops vs GitHub Actions:** Once the MolSSI CMS Cookiecutter is [updated to use GitHub Actions](https://github.com/MolSSI/cookiecutter-cms/pull/95), it would be useful to migrate from Azure Devops to GitHub Actions so that it is easier for folks to just deeply understand one continuous integration system.
* **TeXLive:** We stopped building the docs a little while ago, but it's great if we can continue to have TeXLive in our images if we can resurrect the docs building in the recipe (or a separate package). If we can't get it to work, we can likely drop TeXLive from our images.
* **BINSTAR_TOKEN:** Since this is active under your Azure account, you'll need to pull the current token from https://anaconda.org/omnia-dev/settings/profile (where you are an owner of `omnia-dev` no). When we change this to GitHub Actions later, we can see if there is some sort of institutional account we can use.
* **Build strings:** I like having a `nocuda` label for variants that do not have CUDA support built in, though there should always be a default package with the latest CUDA (if available) on `main`. There's also no point in `nocuda` for `osx` if we're not building CUDA variants.
* **Windows builds:** I am super excited to have CUDA builds for Windows!
* **Updating with conda-forge:** There are a mess of `~/.conda_configs/` files that appear to have been generated with the conda-forge tooling, but how do we update these going forward to stay in sync with the conda-forge infrastructure? Any thoughts on how we make this sustainable?
* **conda-build-all:** Is this script something we should make external, or try to harmonize with the [`build_all.py`](https://github.com/conda-forge/staged-recipes/blob/master/.ci_support/build_all.py) from conda-forge, or the [CI build infrastructure](https://github.com/conda-forge/staged-recipes/tree/master/.ci_support) in its entirety?

@jaimergp will work with you on updating the `openmm` recipe to bring it into alignment with the conda-forge recipe in the areas where this is needed.

Overall, this is great, and we can merge and move forward with iterative improvements!

**Comment by peastman** [comment]
> conda-forge currently officially supports Python 2.7, 3.6 and 3.7, so we should try to support these if possible.

The official end of support date for Python 2.7 is in less than a month, so I don't think it's worth putting in any effort to try to support it.  Also, the next OpenMM release will not support Python 2.

**Comment by jaimergp** [comment]
I'll leave this resource here for future reference. IBM provides packages for `cudatoolkit` and `cudatoolkit-dev` for both `linux64` and `ppc64le`. We can't use these in `conda-forge` because this is an external channel, but might be useful for `omnia` builds at some point?

https://public.dhe.ibm.com/ibmdl/export/pub/software/server/ibm-ai/conda/#/

More info: https://github.com/conda-forge/openmm-feedstock/issues/8#issuecomment-549890209

**Comment by jaimergp** [comment]
One more thing, maybe for a separate PR: PowerPC builds. Nvidia provides base images [here](https://hub.docker.com/r/nvidia/cuda-ppc64le). Do we want to include these builds here too?

**Comment by Lnaden** [comment]
>  currently officially supports Python 2.7, 3.6 and 3.7,

Yes, but Python 2.7 dies on Jan 1 2020, by the time we get all this hammered out, we may as well just not support it. They are also working on phasing out 3.6 and once 3.8 is stable on their builds, 3.6 will drop. I would not say they are supporting 3.6 as much as they are phasing it out. They could drop it any day as far as I know.

> Azure Devops vs GitHub Actions

That is something we can hash after getting the build system upgraded first, one thing at a time.

> TeXLive

I would love to drop this, it adds a fair amount of complexity to the image since I have to build a parallel glibc for the newer (2018+) TeXLives to support.

> BINSTAR_TOKEN: Since this is active under your Azure account,

Its under my azure account so I could test without spamming everyone while I tried a bunch of things and in case it did not work. My plan was to merge this in, and then create a group Azure account which all Omnia devs have full control over. So hopefully we can resolve this without having to move the binstar token between owners.

> Build strings: I like having a nocuda label for variants that do not have CUDA support built in, though there should always be a default package with the latest CUDA (if available) on main. There's also no point in nocuda for osx if we're not building CUDA variants.

This is something I'm not sure how to best handle. I don't know how to make it so when someone does `conda install openmm` it just "gets" the right package with CUDA support. Also, the more complexity we add in the logic for the build string, the harder it is to program. Right now I have Jinja header for setting the CUDA string, I could add another, but it would be good to hammer out exactly what we want the string to be and under what conditions. I do think we should start adding in the `PKG_HASH` to packages going forward for all omnia packages. (see https://github.com/omnia-md/conda-dev-recipes/pull/177#issuecomment-558818411)

> Updating with conda-forge

There are a bunch of them, but I wrote most of them. They are just simple YAML dicts with 1 or 2 options. I have a generator script I have added in a recent change locally I will be pushing.

> conda-build-all...

I would be very, very careful on how much extra work to put into this. I tried to make this so that it pays the technical debt we have accrued over the years while requiring minimal maintenance in the future (hence why I try to use the Conda Build API everywhere I can). Staged recipes is a very special repo from Conda-Forge and I am very weary of trying to *mimic* conda-forge's build system because it makes alot of assumptions we don't (channel rigidity, build images, etc) and has a bunch of active devs which we cannot commit in equal parts.

Overall, I am going to make some changes on my end, and work with @jaimergp to get OpenMM building on my local builds (to at least ensure it works), then we can move forward with merging, tweaking, and fixing. Sound good?

**Comment by jchodera** [comment]
@lnaden @jaimergp: Happy New Year, and welcome back from the holidays!

We desperately need to complete the refurbishment of our omnia build infrastructure, get nightly builds working again, and release the conda-forge version of OpenMM. What else needs to be done here, and is there any way I can help?

**Comment by jchodera** [comment]
(Notably, we haven't pushed an [`omnia-dev` nightly build of `openmm` for two months now](https://anaconda.org/omnia-dev/openmm/files), and we've reached a point where we're going to need the new nightlies once https://github.com/openmm/openmm/pull/2511 is merged until we can get that into a release.)

**Comment by jaimergp** [comment]
I am happy to take over whenever @Lnaden gives the OK!

**Comment by Lnaden** [comment]
I think it might be best to hand it off. I'm not sure when I can finish this in the short term. @jaimergp do you want to have another chat later today or tomorrow afternoon and we can go over all the changes/mechanisms this will introduce as well as what needs done to finish it?

**Comment by Lnaden** [comment]
@jaimergp, I have made some last changes to this, I think its ready to be merged in. It will merge to conda-build3-v2 for a branch instead of master, so you should be able to take control of it from there. I am getting the Azure DevOps org set up now

**Comment by jaimergp** [comment]
@Lnaden LGTM!

**Comment by jaimergp** [comment]
I might need admin access to this org too.

**Comment by Lnaden** [comment]
@jaimergp I have added you to the maintainers for this repo, you should be able to to merging and whatnot after accepting!

**Comment by jaimergp** [comment]
Do we need approval from Peter too or can I just merge this to the dev branch?

**Comment by Lnaden** [comment]
Na, you can just merge this since its going to a branch called `conda-build3-v2` and not master

**Comment by Lnaden** [comment]
Awesome

**Comment by jchodera** [comment]
> Do we need approval from Peter too or can I just merge this to the dev branch?

FYI, Peter hasn't been much involved in the conda build infrastructure, but has generously offered to help maintain it going forward.
